RATIS-2251. Migrate ratis-test tests to Junit 5 - Part 3.

slfan1989 opened the pull request:
## What changes were proposed in this pull request?

Trying to upgrade ratis-server's Junit4 unit tests to Junit5.

## What is the link to the Apache JIRA

JIRA: RATIS-2251. Migrate ratis-test tests to Junit 5 - Part 3.

## How was this patch tested?

Junit Test & mvn clean test.

szetszwo commented on ratis-server/src/test/java/org/apache/ratis/TestReConfigProperty.java:
```diff
@@ -196,13 +197,13 @@ public void testReconfigure() {
       }
-      Assert.assertFalse("received unexpected exception",
-          exceptionCaught);
+      Assertions.assertFalse(
+         exceptionCaught, "received unexpected exception");
```
Let's use `fail(..)` inside the catch-block instead of `assertFalse(..)`.

slfan1989 replied:
@szetszwo Thank you very much for reviewing this PR and providing suggestions! I will continue to improve this PR.

slfan1989 replied:
I need some time to validate the unit tests for the issue and will provide a response as soon as possible.

szetszwo reviewed (commented):
@slfan1989 , thanks a lot for working on this!  The change looks good in general.  Let's also clean up `TestReConfigProperty`; see https://issues.apache.org/jira/secure/attachment/13074931/1227_review.patch

BTW, there are test failures.  Please take a look.

szetszwo commented:
Not sure why it has IllegalStateException in https://github.com/apache/ratis/actions/runs/13490213066/job/37726235717?pr=1227#step:11:886
```
Warning:  Could not process start event for test [JUnit Jupiter executed by VM 1]
java.lang.IllegalStateException: Overlapping ID assignment for 
'HashableTestDescriptor{testDescriptor=DefaultTestDescriptor{id=5866381821535424172, 
surefireForkId=1, name='JUnit Jupiter', className='null', parentId=null, displayName='null'}, 
testMojoExecutionId=-6165802485913919921}' (prev: '3968948658746052783').
```

slfan1989 commented:
> Not sure why it has IllegalStateException in https://github.com/apache/ratis/actions/runs/13490213066/job/37726235717?pr=1227#step:11:886

@szetszwo Thank you for your message. I am currently conducting local tests to identify the cause of the issue.

szetszwo commented:
@slfan1989 , Thanks for the update!

The subclass tests related to `RaftAsyncTests` and `InstallSnapshotNotificationTests` keep failing.  How about  reverting these two files for now and work on them later?

slfan1989 commented:
@szetszwo Thank you very much for your message! I will attempt to roll back these two classes and continue testing locally. I have upgraded the JUnit dependency and the version of the surefire-plugin. The previous java.lang.IllegalStateException error is no longer occurring, and I will continue testing today.

slfan1989 commented:
I have carefully reviewed the CI report, and these errors are likely unrelated to the JUnit 5 upgrade. The same errors existed in JUnit 4 as well and can still be reproduced locally.

Let me summarize the current situation regarding the unit test errors.

> org.apache.ratis.grpc.TestWatchRequestWithGrpc#testWatchRequestAsyncChangeLeader

CI:
```
Error:  Tests run: 5, Failures: 0, Errors: 1, Skipped: 4, Time elapsed: 163.401 s <<< FAILURE! - in org.apache.ratis.grpc.TestWatchRequestWithGrpc
Error:  org.apache.ratis.grpc.TestWatchRequestWithGrpc.testWatchRequestAsyncChangeLeader  Time elapsed: 163.266 s  <<< ERROR!
java.util.concurrent.TimeoutException: testWatchRequestAsyncChangeLeader() timed out after 100 seconds
```

LOCAL JUNIT4 SUCCESS
LOCAL JUNIT5 SUCCESS

> TestInstallSnapshotNotificationWithGrpc#testAddNewFollowersNoSnapshot

CI: 
```
[INFO] Running org.apache.ratis.grpc.TestInstallSnapshotNotificationWithGrpc
Error:  Tests run: 6, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 50.061 s <<< FAILURE! - in org.apache.ratis.grpc.TestInstallSnapshotNotificationWithGrpc
Error:  org.apache.ratis.grpc.TestInstallSnapshotNotificationWithGrpc.testAddNewFollowersNoSnapshot  Time elapsed: 5.588 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <-1> but was: <16>
	at org.apache.ratis.InstallSnapshotNotificationTests.testAddNewFollowers(InstallSnapshotNotificationTests.java:256)
	at org.apache.ratis.InstallSnapshotNotificationTests.lambda$testAddNewFollowersNoSnapshot$1(InstallSnapshotNotificationTests.java:182)
	at org.apache.ratis.server.impl.MiniRaftCluster$Factory$Get.runWithNewCluster(MiniRaftCluster.java:144)
	at org.apache.ratis.server.impl.MiniRaftCluster$Factory$Get.runWithNewCluster(MiniRaftCluster.java:121)
	at org.apache.ratis.InstallSnapshotNotificationTests.testAddNewFollowersNoSnapshot(InstallSnapshotNotificationTests.java:182)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
```

Local:
```
2025-03-02 21:50:06,575 [Time-limited test] INFO  impl.MiniRaftCluster (MiniRaftCluster.java:shutdown(883)) - MiniRaftClusterWithGrpc shutdown completed
2025-03-02 21:50:06,576 [Time-limited test] INFO  ratis.RaftTestUtil (RaftTestUtil.java:gc(596)) - gc 0
2025-03-02 21:50:06,705 [Time-limited test] INFO  impl.MiniRaftCluster (MiniRaftClusterWithGrpc.java:lambda$null$2(103)) - Checking s0-group-96D8828789EC
2025-03-02 21:50:06,706 [Time-limited test] INFO  impl.MiniRaftCluster (MiniRaftClusterWithGrpc.java:lambda$null$2(103)) - Checking s1-group-96D8828789EC
2025-03-02 21:50:06,706 [Time-limited test] INFO  impl.MiniRaftCluster (MiniRaftClusterWithGrpc.java:lambda$null$2(103)) - Checking s2-group-96D8828789EC

java.lang.AssertionError: 
Expected :-1
Actual   :16
<Click to see difference>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at org.apache.ratis.InstallSnapshotNotificationTests.testAddNewFollowers(InstallSnapshotNotificationTests.java:256)
	at org.apache.ratis.InstallSnapshotNotificationTests.lambda$testAddNewFollowersNoSnapshot$1(InstallSnapshotNotificationTests.java:182)
	at org.apache.ratis.server.impl.MiniRaftCluster$Factory$Get.runWithNewCluster(MiniRaftCluster.java:144)
	at org.apache.ratis.server.impl.MiniRaftCluster$Factory$Get.runWithNewCluster(MiniRaftCluster.java:121)
	at org.apache.ratis.InstallSnapshotNotificationTests.testAddNewFollowersNoSnapshot(InstallSnapshotNotificationTests.java:182)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

> TestRaftAsyncWithNetty

CI:
```
[INFO] Running org.apache.ratis.netty.TestRaftAsyncWithNetty
Error:  Tests run: 14, Failures: 1, Errors: 7, Skipped: 0, Time elapsed: 348.303 s <<< FAILURE! - in Error:  

# 1 TestRaftAsyncWithNetty#testStaleReadAsync
Error:  org.apache.ratis.netty.TestRaftAsyncWithNetty.testStaleReadAsync  Time elapsed: 6.909 s  <<< ERROR!
java.util.concurrent.CompletionException: org.apache.ratis.protocol.exceptions.NotLeaderException: Server s0@group-CC3D2906FD90 is not the leader, suggested leader is: s1|localhost:15004
	at org.apache.ratis.client.impl.RaftClientImpl.handleRaftException(RaftClientImpl.java:373)
	at org.apache.ratis.client.impl.OrderedAsync.lambda$send$3(OrderedAsync.java:175)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:616)

# 2 TestRaftAsyncWithNetty.#estStateMachineMetrics
Error:  org.apache.ratis.netty.TestRaftAsyncWithNetty.testStateMachineMetrics  Time elapsed: 1.417 s  <<< ERROR!
java.util.concurrent.ExecutionException: org.apache.ratis.protocol.exceptions.NotLeaderException: Server s0@group-DA17DD87FDE1 is not the leader, suggested leader is: s1|localhost:15016
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)

# 3 TestRaftAsyncWithNetty.testAppendEntriesTimeout
Error:  org.apache.ratis.netty.TestRaftAsyncWithNetty.testAppendEntriesTimeout  Time elapsed: 6.356 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <false> but was: <true>
	at org.apache.ratis.RaftAsyncTests.runTestAppendEntriesTimeout(RaftAsyncTests.java:433)
	at org.apache.ratis.server.impl.MiniRaftCluster$Factory$Get.runWithNewCluster(MiniRaftCluster.java:144)
	at org.apache.ratis.server.impl.MiniRaftCluster$Factory$Get.runWithNewCluster(MiniRaftCluster.java:121)
	at org.apache.ratis.RaftAsyncTests.testAppendEntriesTimeout(RaftAsyncTests.java:413)
	at java.lang.reflect.Method.invoke(Method.java:498)

# 4 TestRaftAsyncWithNetty.testWithLoadAsync
Error:  org.apache.ratis.netty.TestRaftAsyncWithNetty.testWithLoadAsync  Time elapsed: 100.973 s  <<< ERROR!
java.util.concurrent.TimeoutException: testWithLoadAsync() timed out after 100 seconds
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	Suppressed: java.lang.InterruptedException: sleep interrupted

# 5 TestRaftAsyncWithNetty.testRequestTimeout
Error:  org.apache.ratis.netty.TestRaftAsyncWithNetty.testRequestTimeout  Time elapsed: 1.395 s  <<< ERROR!
java.util.concurrent.ExecutionException: org.apache.ratis.protocol.exceptions.NotLeaderException: Server s0@group-2B35970A0B90 is not the leader, suggested leader is: s1|localhost:15088
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)
	at org.apache.ratis.RaftBasicTests.testRequestTimeout(RaftBasicTests.java:438)

# 6 TestRaftAsyncWithNetty#testAsyncRequestSemaphore
Error:  org.apache.ratis.netty.TestRaftAsyncWithNetty.testAsyncRequestSemaphore  Time elapsed: 100.379 s  <<< ERROR!
java.util.concurrent.TimeoutException: testAsyncRequestSemaphore() timed out after 100 seconds
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at java.util.ArrayList.forEach(ArrayList.java:1259)

# 7 TestRaftAsyncWithNetty#testNoRetryWaitOnNotLeaderException
Error:  org.apache.ratis.netty.TestRaftAsyncWithNetty.testNoRetryWaitOnNotLeaderException  Time elapsed: 1.636 s  <<< ERROR!
java.util.concurrent.ExecutionException: org.apache.ratis.protocol.exceptions.NotLeaderException: Server s1@group-8CF8D02C8046 is not the leader, suggested leader is: s0|localhost:15120
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1928)
	at org.apache.ratis.util.TimeDuration.apply(TimeDuration.java:313)

# 8 TestRaftAsyncWithNetty#testRequestAsyncWithRetryFailureAfterInitialMessages
Error:  org.apache.ratis.netty.TestRaftAsyncWithNetty.testRequestAsyncWithRetryFailureAfterInitialMessages  Time elapsed: 100.97 s  <<< ERROR!
java.util.concurrent.TimeoutException: testRequestAsyncWithRetryFailureAfterInitialMessages() timed out after 100 seconds
	at java.util.ArrayList.forEach(ArrayList.java:1259)
```

LOCAL:

![image](https://github.com/user-attachments/assets/534b326e-9822-45f7-aa63-554f3f8f3b0d)

szetszwo commented:
@slfan1989 , after RATIS-2124 is merged, we need to resolve the conflicts here.  Thanks.

slfan1989 commented:
> @slfan1989 , after [RATIS-2124](https://issues.apache.org/jira/browse/RATIS-2124) is merged, we need to resolve the conflicts here. Thanks.

@szetszwo Thank you for your message! I will continue to follow up on this PR.

szetszwo commented on pom.xml:
```diff
@@ -174,7 +174,7 @@
     <maven-remote-resources-plugin.version>3.3.0</maven-remote-resources-plugin.version>
     <maven-shade-plugin.version>3.6.0</maven-shade-plugin.version>
-    <maven-surefire-plugin.version>3.0.0-M4</maven-surefire-plugin.version>
+    <maven-surefire-plugin.version>3.0.0-M9</maven-surefire-plugin.version>
```
How about try updating it to the latest 3.5.2?
https://maven.apache.org/surefire/maven-surefire-plugin/usage.html

adoroszlai replied:
I tried updating surefire to 3.5.2, but had some problems:

```
Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.5.2:test (default-test) on project ratis-docs: groups/excludedGroups require TestNG, JUnit48+ or JUnit 5 (a specific engine required on classpath) on project test classpath
Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.5.2:test (default-test) on project ratis-proto: groups/excludedGroups require TestNG, JUnit48+ or JUnit 5 (a specific engine required on classpath) on project test classpath
```

These modules have `skipTests=true`, so I don't understand why surefire wants to run tests.  Also tried with `maven.test.skip=true`, which should even skip test compilation, but the problem still happened.

It would be great if you can make it work.

https://github.com/adoroszlai/ratis/commit/d53a586d3ccae780c152841296e5df9314d4113d
https://github.com/adoroszlai/ratis/actions/runs/13670191390

szetszwo replied:
@adoroszlai , thanks for checking!  It succeeded when I run it locally but filed in GitHub https://github.com/szetszwo/ratis/actions/runs/13725256050

It may be a problem in our githut action configuration.

adoroszlai replied:
The error says it's related to using `excludedGroups`:

```
groups/excludedGroups require TestNG, JUnit48+ or JUnit 5 (a specific engine required on classpath) on project test classpath
```

which is set in CI:

https://github.com/apache/ratis/blob/d740e513555a3c4bbf26cab061bca5ff63bb1859/pom.xml#L1122

This should not happen with [`skipTests`](https://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html#skiptests), which we set for `ratis-docs` here:

https://github.com/apache/ratis/blob/d740e513555a3c4bbf26cab061bca5ff63bb1859/ratis-docs/pom.xml#L31-L32

szetszwo replied:
Interestingly, after added `set -x`, it can pass all the checks; see https://github.com/szetszwo/ratis/actions/runs/13776351863

szetszwo replied:
Added `echo` to print out the command
```
mvnw -V -B -Dsurefire.rerunFailingTestsCount=5 -Dsurefire.timeout=1200 --fail-at-end -Djacoco.skip test -Pflaky-tests
```
Now, I can reproduce it locally using the above command.
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  13.154 s
[INFO] Finished at: 2025-03-10T17:51:38-07:00
[INFO] ------------------------------------------------------------------------
[INFO] 42 goals, 42 executed
[INFO] Build cache (/Users/szetszwo/.m2/.develocity/build-cache/v1) removing files not accessed on or after Mon Mar 03 16:51:38 PST 2025.
[INFO] Build cache (/Users/szetszwo/.m2/.develocity/build-cache/v1) cleanup deleted 0 files/directories.
[INFO] Build cache (/Users/szetszwo/.m2/.develocity/build-cache/v1) cleaned up in 0.003 secs.
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.5.2:test (default-test) on project ratis-docs: groups/excludedGroups require TestNG, JUnit48+ or JUnit 5 (a specific engine required on classpath) on project test classpath -> [Help 1]
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.5.2:test (default-test) on project ratis-proto: groups/excludedGroups require TestNG, JUnit48+ or JUnit 5 (a specific engine required on classpath) on project test classpath -> [Help 1]
```

slfan1989 replied:
@szetszwo @adoroszlai Thank you for the discussion! it has been very helpful to me. However, I would like to share my personal thoughts. Based on my own debugging, I confirm that there are issues with the current unit tests. Whether or not JUnit 5 is used, errors still occur.

I haven’t debugged all the unit tests, but I did focus on TestInstallSnapshotNotificationWithGrpc.testAddNewFollowersNoSnapshot. This unit test seems to have been affected by code changes in subsequent commits, which led to the error. When I rolled back to a much earlier version, I found that this test passed without issues.

Additionally, thank you for providing https://github.com/apache/ratis/actions/runs/13530776533/job/37812219666?pr=1228#step:11:923. 

However, in the test report, we noticed that `TestInstallSnapshotNotificationWithGrpc` didn’t actually run. It seems that after `TestRaftWithGrpc`, the unit tests were interrupted.

I plan to submit a PR that triggers unit tests. We can use this PR to confirm whether the unit tests are working properly.

The upgrade of the Surefire plugin is indeed challenging. In the Hadoop project, we attempted to upgrade twice but encountered some issues both times(Many unit tests were skipped.). Therefore, I think it’s best to handle this separately, as doing everything at once could make things more complicated.

My suggestion is that if a PR triggering unit tests shows errors in the same class, we can temporarily disable the problematic tests and continue with the JUnit 5 upgrade. After that, we can address these failing tests one by one. Is this approach acceptable?

adoroszlai replied:
Please do not disable tests.

1. Trigger [repeated run](https://github.com/slfan1989/ratis/actions/workflows/repeat-test.yaml) to confirm flakiness and establish failure rate.
2. Create Jira issue to record the problem.
3. Mark test as `@Flaky`, referencing the Jira issue.

`@Flaky` tests are repeated automatically in CI.

Note, it does not help for tests failing due to leaks, as state of leak detector is not reset between retries.

slfan1989 replied:
I agree with your suggestion, and I will continue to follow up on solving the unit testing issue.

szetszwo reviewed (commented):
@slfan1989 , thanks a lot for you hard work!  Not sure why the unit tests keep failing here but not the other PRs such as https://github.com/apache/ratis/pull/1228 .  Let's try updating surefire to see if it make a difference.

slfan1989 commented:
Apologies for not effectively following up on this PR. I will now continue to follow up on it.

slfan1989 commented:
@szetszwo @adoroszlai 

When addressing the issue with the unit test : `TestInstallSnapshotNotificationWithGrpc#testAddNewFollowersNoSnapshot`

I found that the error persists even after applying the fix from RATIS-2045. 

RATIS-2045 fixed the issue where SnapshotInstallationHandler didn't notify followers to install snapshots when the snapshot index was -1 and the leader's firstAvailableLogIndex was 0 (PR #1053).

This PR changes the behavior of whether followers pull snapshots from the leader.

From the logs, we can observe that the newly added followers `s1` and `s2` have both synchronized snapshots from the leader `s0`. As a result, the snapshot index for followers `s1` and `s2` becomes `16` (16 because we manually created messages twice), instead of `-1`. Therefore, the current check condition is problematic.

```
follower s1:
2025-05-10 17:23:10,134 [grpc-default-executor-2] INFO  impl.SnapshotInstallationHandler (SnapshotInstallationHandler.java:notifyStateMachineToInstallSnapshot(262)) - s1@group-F83BA0BDB609: Received notification to install snapshot at index 0
2025-05-10 17:23:10,137 [grpc-default-executor-2] INFO  impl.SnapshotInstallationHandler (SnapshotInstallationHandler.java:notifyStateMachineToInstallSnapshot(297)) - s1@group-F83BA0BDB609: notifyInstallSnapshot: nextIndex is 0 but the leader's first available index is 0.
......
2025-05-10 17:23:11,151 [grpc-default-executor-0] INFO  impl.SnapshotInstallationHandler (SnapshotInstallationHandler.java:notifyStateMachineToInstallSnapshot(365)) - s1@group-F83BA0BDB609: InstallSnapshot notification result: SNAPSHOT_INSTALLED, at index: (t:1, i:16)

follower s2:
2025-05-10 17:23:11,214 [grpc-default-executor-2] INFO  impl.SnapshotInstallationHandler (SnapshotInstallationHandler.java:notifyStateMachineToInstallSnapshot(262)) - s2@group-F83BA0BDB609: Received notification to install snapshot at index 0
2025-05-10 17:23:11,214 [grpc-default-executor-2] INFO  impl.SnapshotInstallationHandler (SnapshotInstallationHandler.java:notifyStateMachineToInstallSnapshot(297)) - s2@group-F83BA0BDB609: notifyInstallSnapshot: nextIndex is 0 but the leader's first available index is 0.
......
2025-05-10 17:23:12,217 [grpc-default-executor-0] INFO  impl.SnapshotInstallationHandler (SnapshotInstallationHandler.java:notifyStateMachineToInstallSnapshot(365)) - s2@group-F83BA0BDB609: InstallSnapshot notification result: SNAPSHOT_INSTALLED, at index: (t:1, i:16)
```

Logs before applying RATIS-2045:

```
2025-05-10 17:42:54,878 [grpc-default-executor-0] INFO  impl.SnapshotInstallationHandler (SnapshotInstallationHandler.java:notifyStateMachineToInstallSnapshot(221)) - s1@group-46FD094EFC86: Received notification to install snapshot at index 0
2025-05-10 17:42:54,878 [grpc-default-executor-0] INFO  impl.SnapshotInstallationHandler (SnapshotInstallationHandler.java:notifyStateMachineToInstallSnapshot(230)) - s1@group-46FD094EFC86: InstallSnapshot notification result: ALREADY_INSTALLED, current snapshot index: -1
.....
2025-05-10 17:42:54,880 [grpc-default-executor-2] INFO  impl.SnapshotInstallationHandler (SnapshotInstallationHandler.java:notifyStateMachineToInstallSnapshot(221)) - s2@group-46FD094EFC86: Received notification to install snapshot at index 0
2025-05-10 17:42:54,880 [grpc-default-executor-2] INFO  impl.SnapshotInstallationHandler (SnapshotInstallationHandler.java:notifyStateMachineToInstallSnapshot(230)) - s2@group-46FD094EFC86: InstallSnapshot notification result: ALREADY_INSTALLED, current snapshot index: -1
```

So, if we believe that #1053 is reasonable, we should modify the check condition by adjusting the expected value to match the leader's value.

![image](https://github.com/user-attachments/assets/a48a864f-6bbf-4c8b-840f-43ca35112453)

Unit Test Result:

![image](https://github.com/user-attachments/assets/ea6ba123-b883-41f6-a209-016da9688dde)

slfan1989 commented:
@szetszwo @adoroszlai 

I’ve locally fixed the issues in the `TestRaftAsyncWithNetty` unit test. Below is an analysis of the problem and its root causes:

First, the test failures are not related to our upgrade to `JUnit5`. The tests also fail under `JUnit4`.

There are two key issues:

`NettyClientRpc` lacks proper exception handling: When an exception occurs, it is thrown directly, which prevents the retry mechanism from being triggered.

`NettyClientRpc` does not implement timeout handling: In timeout scenarios, there is no effective control or handling logic in place. 

The test results are as follows:

<img width="1429" alt="image" src="https://github.com/user-attachments/assets/d90d9097-023e-4bd9-90b0-e12ed054975b" />

- Enhanced exception handling logic in `ratis-netty`

After comparing the execution flow of `ratis-grpc`, I found that it handles exceptions properly, which allows the retry mechanism at the `ratis-rpc` layer to be triggered. However, `ratis-netty` lacks this corresponding logic.

This leads to an issue: in the `ratis-netty` unit test (`TestRaftAsyncWithNetty#testStaleReadAsync`), the test fails immediately if `s0` is not the leader. In contrast, `ratis-grpc` handles the `NotLeaderException` properly, which triggers the retry mechanism, allowing the test to pass successfully.

Exception handling logic in `ratis-grpc`

https://github.com/apache/ratis/blob/2eda35dd532f503b299db9e3263ecbeca52da023/ratis-grpc/src/main/java/org/apache/ratis/grpc/client/GrpcClientProtocolClient.java#L307-L329

I implemented similar logic in the `NettyClientRpc.java` code.

- Added timeout handling logic  in `ratis-netty`

The unit test(`misc`) for RATIS-2251 frequently times out. The original timeout was set to 60 minutes, and even after increasing it to 90 and 120 minutes, the test still timed out. This indicates that the issue likely lies within the test itself, rather than the configured timeout duration.

The root cause is that `ratis-netty` lacks a timeout mechanism, which causes the unit test to hang for extended periods when exceptions occur or there is no response.

`ratis-grpc` includes corresponding exception and timeout handling logic, as shown below:

https://github.com/apache/ratis/blob/2eda35dd532f503b299db9e3263ecbeca52da023/ratis-grpc/src/main/java/org/apache/ratis/grpc/client/GrpcClientProtocolClient.java#L367-L371

I implemented similar logic in the `NettyClientRpc.java` code.

slfan1989 commented:
@szetszwo @adoroszlai RATIS-2251 has passed all checks. During repeated testing, I identified a new flaky test: `RaftReconfigurationBaseTest#testKillLeaderDuringReconf`, and I have added the appropriate unit test annotation for it.
Could you please help review this PR again? Thank you very much!

szetszwo commented on ratis-server/src/test/java/org/apache/ratis/InstallSnapshotNotificationTests.java:
```diff
@@ -220,7 +220,7 @@ private void testAddNewFollowers(CLUSTER cluster, int numRequests) throws Except
     LOG.info("Delete logs {}", logs);
     for (LogSegmentPath path : logs) {
-      FileUtils.deleteFully(path.getPath()); // the log may be already purged
+      FileUtils.deleteFully(path.getPath()); // the log may be already puged
```
I guess you might have accidentally edited the comment.

slfan1989 replied:
Thank you very much for reviewing the code! This issue was indeed an oversight on my part. I will fix it along with the flaky tests in the upcoming revisions.

szetszwo reviewed (commented):
Just found a typo in a java comment.  Since it is very minor, let just merge this.

+1 the change looks good.
